cjworkbench/i18n/trans.py

```diff
@@ -53,16 +53,19 @@ def trans(
 """
 
 
-def localize(
-    locale_id: str, message_id: str, parameters: _Parameters = {}
-) -> Optional[str]:
+def localize(locale_id: str, message_id: str, parameters: _Parameters = {}) -> str:
     """Localize the given message ID to the given locale.
 
-    If the message is not found in the catalog for the given or the default locale, return `None`.
+    If the message is not found in the catalog for the given or the default locale, return `None`,
+    raise `KeyError`.
     """
-    return _get_translations(locale_id).trans(
+    message = _get_translations(locale_id).trans(
```

Comment:
This changes the behavior we had before. I feel the previous behavior was better. Thoughts?

Previously, `trans()` would:

1. Try the user-locale message catalog. If there's an ICUError, log the error and fall through.
2. Try the default-locale message catalog. If there's an ICUError, raise it.

After applying this pull request, we get:

1. Try the user-locale message catalog. If there's an ICUError, log the error and fall through.
2. Try the default-locale message catalog. If there's an ICUError, log the error and fall through.
3. Raise KeyError.

I think the simplest implementation is to rip out all the `Optional` from `MessageTranslator`. `MessageTranslator.get_message()` would raise `KeyError` if there's an error; `MessageTranslator.trans()` would raise `KeyError` or `ICUError`. And `localize()` would handle them all, just as `MessageTranslator.trans()` used to.

Also, would it be a good idea to name it `MessageTranslator.get_and_format_message()` instead of `MessageTranslator.trans()`? The name itself would make it obvious which errors might be raised.

Reply:
Part of this behaviour is new, in the sense that `trans` used its `default` argument as a last resort and we want to stop doing that. Hence, if the message is nowhere to be found, `localize` (and `trans`) raises the `KeyError` without handling it, correct?

Reply:
Correct.

The `KeyError` is new, yes ... but that shouldn't affect `trans()` in practice. `trans()` can only raise `KeyError` if a developer fails to `bin/dev extract-messages` -- in which case, we _want_ a crash. Aside from that one case, I like the `trans()` behavior on master (its throwing of `ICUError`, for instance); so let's keep it.

In contrast, `localize()` is new. And the more I think about it, the more I feel it shouldn't handle any exceptions at all. Imagine:

```
def localize(locale_id: str, message_id: str, parameters: _Parameters) -> str:
    translations = _get_translations(locale_id)
    return translations.get_and_format_message(message_id, parameters)  # raise KeyError, ICUError
```

No need for default arguments. The only callers are `trans()` and `jsonize_i18n_message()`, and they both pass arguments. (Also, nit: this should be named "arguments", not "parameters", as ICU MessageFormat terminology is "arguments".)

No need for exception handlers. Callers handle exceptions differently.

No need for `Optional`. It's just added confusion.

It's easy to envision how `trans()` would work with this. And it's easy to envision how `jsonize_i18n_message()` would work.

Here's my proposal:

* `trans()` is a general-purpose function. It is used within Workbench to log Workbench messages. It crashes if the _developer_ made a mistake -- either forgot to extract messages or supplied an invalid message format.
* `localize()` is a low-level utility function.
* `jsonize_i18n_message()` is a high-level function specifically designed to handle I18nMessages from modules. Those messages are untrusted, so the tasks are A) log errors so module authors can see them; and B) recover from errors so output is always sane.

Reply:
The only thing missing from your comment is that `localize`, although low-level, has the job of implementing the fallback mechanism to the default locale. Hence, it needs to handle the errors of missing or badly formatted messages in the message catalogs of non-default locales.

For completeness, I should mention that with this pull request `trans` will also crash if the developer edits the English po file directly and makes a message invalid: it will crash when trying to use the message.

Reply:
Well, in my previous comment I was suggesting a bit of a different direction. I was suggesting `localize()` should do less, `trans()` and `jsonize_i18n_message()` should pick up the slack. But that doesn't matter right now -- your new edits to this pull request cover all today's scenarios, so it's worth merging as-is. Later, when we implement multiple sources, we'll adjust.